compiler/src/iree/compiler/Utils/CommonTypeConstraints.td

```diff
@@ -28,4 +28,9 @@ class RankedTensorOrVectorOf<list<Type> allowedTypes> :
 
 def AnyRankedTensorOrVector : RankedTensorOrVectorOf<[AnyType]>;
 
+class RankedTensorOrVectorOrMemRefOf<list<Type> allowedTypes> :
+  ShapedContainerType<allowedTypes,
+      Or<[IsVectorOfAnyRankTypePred, IsTensorTypePred, IsMemRefTypePred]>,
```

Comment:
IsTensorTypePred does not require that it's ranked, you need to do `And<[IsTensorTypePred, HasRankPred]>` (seen a few lines up).